Update performance budgets tooling guide (#1104)
GoogleChrome/web.dev: src/site/content/en/fast/incorporate-performance-budgets-into-your-build-tools/index.md

Once you’ve defined a performance budget, it’s time to set up the build process
to keep track of it. There are a number of tools that let you define thresholds
for chosen performance metrics and warn you if you go over budget. Find out
how to choose one that best fits your needs and current setup. 🕵️‍♀️

## Lighthouse performance budgets

[Lighthouse](https://developers.google.com/web/tools/lighthouse/) is an auditing tool that tests sites in a few key areas — performance, accessibility, best practices and how well your site performs as a progressive web application.

The [command line version](https://developers.google.com/web/tools/lighthouse/#cli) of Lighthouse (v5+) supports setting [performance budgets](https://developers.google.com/web/tools/lighthouse/audits/budgets) based on:

- resource size
- resource count

You can set budgets for any of the following resource types: 
- `document` 
- `font`
- `image`
- `media`
- `other`
- `script`
- `stylesheet`
- `third-party`
- `total`

> Budgets are set in a JSON file and after defining them the new Over Budget column tells you whether you're exceeding any limits.

Reviewer: ```suggestion
Budgets are set in a JSON file and after defining them the new "Over Budget" column tells you whether you're exceeding any limits.
```

<figure class="w-figure">

>   <img src="./lightwallet.png" alt="Lighthouse Bot">

Reviewer: Broken image, `alt` probably also needs to be updated.

Author: This is also not supposed to be broken, [here's the image](https://github.com/GoogleChrome/web.dev/blob/perf-budgets/src/site/content/en/fast/incorporate-performance-budgets-into-your-build-tools/lightwallet.png).

Will updated alt when I figure out where are my images 👍

  <figcaption class="w-figcaption">
  </figcaption>
</figure>

## Webpack performance hints

[Webpack](https://developers.google.com/web/fundamentals/performance/webpack/) is a powerful build tool for optimizing how your code is delivered to the users. It also supports setting performance budgets based on **asset size**.

> Turn on [performance hints](https://webpack.js.org/configuration/performance/) in the configuration file to get command line warnings or errors when your bundle size grows over the limit. It’s a great way to stay mindful about asset sizes throughout the development.

Reviewer: ```suggestion
Turn on [performance hints](https://webpack.js.org/configuration/performance/) in `webpack.config.js` to get command line warnings or errors when your bundle size grows over the limit. It’s a great way to stay mindful about asset sizes throughout the development.
```

  </figcaption>
</figure>

> The numbers are compared against **uncompressed asset sizes**. This is not an ideal situation, since most hosting platforms, CDNs and reverse proxy servers compress assets by default. You can give yourself some wiggle room during development, but keep in mind that compression speeds up only the transfer. Browsers still have to execute uncompressed files and the [cost of this is not small](https://v8.dev/blog/cost-of-javascript-2019), especially on mobile devices.

Reviewer: ```suggestion
Budgets are compared against **uncompressed asset sizes**. This is not an ideal situation, since most hosting platforms, CDNs and reverse proxy servers compress assets by default. You can give yourself some wiggle room during development, but keep in mind that compression speeds up only the transfer. Browsers still have to execute uncompressed files and the [cost of this is not small](https://v8.dev/blog/cost-of-javascript-2019), especially on mobile devices.
```

Reviewer: I think I'd rework this paragraph a bit. I don't think comparing again uncompressed asset sizes is necessarily bad - it's just that compressed asset sizes typically correlate with transfer times; while uncompressed asset sizes correlate with parse/compile/execution times.

Author: We're not really saying it's bad, it's "not ideal" and we acknowledge that uncompressed sizes affect execution time. How about:

> "You can give yourself some wiggle room during development, but that doesn't mean setting lax budgets and counting on compression to kick in. Keep in mind that compression speeds up only the transfer and browsers still have to execute uncompressed files..."

Reviewer: Ah, so I’d probably argue that using uncompressed asset sizes really isn’t that bad and is actually totally fine.

There’s a ton of variation, so it really depends on a particular person and their device + network, but generally parse/compile/execute is now a bigger issue than network times, which would correlate more with uncompressed sizes. However, historically, networks have been the biggest issue, which I think is why there’s still a tendency for budgets and tooling to be based around transfer size rather than “actual” size. Neither metric (compressed/uncompressed) is perfect, but both have their uses.

Author: This was discussed when I was first writing this guide. At the time, the consensus was that, because our own tooling and budgeting guidelines are based on compressed sizes, we want to echo that preference here. Pretty sure there was some talk about nudging webpack to adopt that approach too.

I'm fine with us being less opinionated, but let's consult @addyosmani ?

Reviewer: SGTM. @addyosmani thoughts on this?

Reviewer 2: I'm also okay with us being less opinionated in the short term. Until we are at a place where there's more consistency across perf budget tooling in terms of whether they're looking at compressed vs uncompressed numbers, this seems like a fair compromise.

  </figcaption>
</figure>

> [Lighthouse Bot](https://github.com/ebidel/lighthouse-ci) integrates with Travis CI and enforces budgets based on any of the five Lighthouse audit categories. It’s sometimes simpler to keep an eye on a single number than individual asset budgets and Lighthouse scores take many things into account.

Reviewer: After the first sentence, I'd add a brief example. (e.g. "For example, you could set a budget of 90 for your Lighthouse performance score.")

<figure class="w-figure">

>   <img class="screenshot" src="./lighthouse-scores.png" alt="Lighthouse scores 💯">

Reviewer: Could you grab a new screenshot of the Lighthouse report? Might as well change it while this article is being updated.

Author: I did that, you should be seeing this: [lighthouse-scores.png](https://github.com/GoogleChrome/web.dev/blob/perf-budgets/src/site/content/en/fast/incorporate-performance-budgets-into-your-build-tools/lighthouse-scores.png)

Not sure what went wrong?

Reviewer: Weird. lighthouse-scores.png still looks like the old version to me.

This is what Lighthouse looks like for me locally:
<img width="895" alt="Screen Shot 2019-07-08 at 1 11 39 PM" src="https://user-images.githubusercontent.com/6076184/60829082-1adca900-a182-11e9-8a6b-314c83a9d72c.png">

Author: Oh you mean *that* new version. My screenshot was from the latest stable Chrome (75), I'll just grab one from the extension 👍

  <figcaption class="w-figcaption">
    Lighthouse scores 💯 
  </figcaption>
</figure>

> Lighthouse Bot runs an audit after you deploy a site to staging server. In `.travis.yml` file set targets with `--perf`, `--a11y`, `--bp`, `--seo` or `--pwa` options. Aim to stay in the green zone with a performance score of at least 90.

Reviewer: ```suggestion
Lighthouse Bot runs an audit after you deploy a site to staging server. In `.travis.yml` set budgets for particular Lighthouse categories by using the `--perf`, `--a11y`, `--bp`, `--seo` or `--pwa` options. Aim to stay in the green zone with a performance score of at least 90.
```

  </figcaption>
</figure>

> If you find your pull request blocked by a poor Lighthouse score, run an audit with [Lighthouse CLI](https://developers.google.com/web/tools/lighthouse/#cli) or in [Dev Tools](https://developers.google.com/web/tools/lighthouse/#devtools). It generates a report with details about bottlenecks and hints for simple optimizations.

Reviewer: ```suggestion
If you find your pull request blocked by a poor Lighthouse score, run an audit with [Lighthouse CLI](https://developers.google.com/web/tools/lighthouse/#cli) or in [Dev Tools](https://developers.google.com/web/tools/lighthouse/#devtools). It generates a report with details about bottlenecks and hints for simple optimizations.
```
```suggestion
If you find your pull request blocked by a poor Lighthouse score, run an audit with [Lighthouse CLI](https://developers.google.com/web/tools/lighthouse/#cli) or in [Dev Tools](https://developers.google.com/web/tools/lighthouse/#devtools). You’ll get a report with details about bottlenecks and hints for simple optimizations.
```

      <th>Tool</th>
      <th>CLI</th>
      <th>CI</th>
      <th>Summary</th>
    </tr>
    <tr>
      <td>Lighthouse</td>
      <td>✅</td>
      <td>✘</td>
      <td>
        <ul>
          <li>Budgets for different types of resources based on their size or count.</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Webpack</td>
      <td>✅</td>
      <td>✘</td>
      <td>
        <ul>
          <li>Budgets based on sizes of assets generated by webpack.</li>
          <li>Checks uncompressed sizes.</li>
        </ul>
      </td>
    </tr>
    <tr>
      <td>Bundlesize</td>
      <td>✅</td>
      <td>✅</td>
      <td>
        <ul>
          <li>Budgets based on sizes of specific resources.</li>

>           <li>Checks compressed sizes.</li>

Reviewer: ```suggestion
          <li>Checks compressed or uncompressed sizes.</li>
```
